File: linux/internal/s3/sigv4.go
package s3

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"
)

// EmptyPayloadSHA256 is SHA256 of the empty string (unsigned GET/DELETE bodies).
const EmptyPayloadSHA256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"

// Credentials for AWS Signature Version 4 (S3).
type Credentials struct {
	AccessKey string
	SecretKey string
	Region    string
	Service   string // default "s3"
}

func (c Credentials) service() string {
	if c.Service == "" {
		return "s3"
	}
	return c.Service
}

// SignRequest adds SigV4 Authorization and related headers to req.
// payloadHash is the hex SHA256 of the body (or EmptyPayloadSHA256 / UNSIGNED-PAYLOAD).
func SignRequest(req *http.Request, creds Credentials, payloadHash string, now time.Time) error {
	if payloadHash == "" {
		payloadHash = EmptyPayloadSHA256
	}
	if now.IsZero() {
		now = time.Now().UTC()
	} else {
		now = now.UTC()
	}

	amzDate := now.Format("20060102T150405Z")
	dateStamp := amzDate[:8]

	hostHeader := req.URL.Host
	if hostHeader == "" {
		hostHeader = req.Host
	}
	req.Header.Set("Host", hostHeader)
	req.Header.Set("X-Amz-Date", amzDate)
	req.Header.Set("X-Amz-Content-Sha256", payloadHash)

	canonicalURI := CanonicalPath(req.URL)
	canonicalQuery := CanonicalQueryString(req.URL)

	// Headers we sign: host, x-amz-content-sha256, x-amz-date (lowercase).
	headers := map[string]string{
		"host":                 hostHeader,
		"x-amz-content-sha256": payloadHash,
		"x-amz-date":           amzDate,
	}
	keys := make([]string, 0, len(headers))
	for k := range headers {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var canonicalHeaders strings.Builder
	for _, k := range keys {
		canonicalHeaders.WriteString(k)
		canonicalHeaders.WriteByte(':')
		canonicalHeaders.WriteString(strings.TrimSpace(headers[k]))
		canonicalHeaders.WriteByte('\n')
	}
	signedHeaders := strings.Join(keys, ";")

	method := req.Method
	if method == "" {
		method = http.MethodGet
	}
	canonicalRequest := strings.Join([]string{
		method,
		canonicalURI,
		canonicalQuery,
		canonicalHeaders.String(),
		signedHeaders,
		payloadHash,
	}, "\n")

	scope := fmt.Sprintf("%s/%s/%s/aws4_request", dateStamp, creds.Region, creds.service())
	stringToSign := strings.Join([]string{
		"AWS4-HMAC-SHA256",
		amzDate,
		scope,
		hexSHA256([]byte(canonicalRequest)),
	}, "\n")

	signingKey := deriveKey(creds.SecretKey, dateStamp, creds.Region, creds.service())
	signature := hex.EncodeToString(hmacSHA256(signingKey, []byte(stringToSign)))
	for i := range signingKey {
		signingKey[i] = 0
	}

	auth := fmt.Sprintf(
		"AWS4-HMAC-SHA256 Credential=%s/%s, SignedHeaders=%s, Signature=%s",
		creds.AccessKey, scope, signedHeaders, signature,
	)
	req.Header.Set("Authorization", auth)
	return nil
}

// CanonicalPath encodes the path for SigV4: segments percent-encoded, "/" preserved.
// Uses EscapedPath (not Path) so a trailing slash on ListObjectsV2 is kept.
func CanonicalPath(u *url.URL) string {
	raw := u.EscapedPath()
	if raw == "" {
		raw = "/"
	}
	decoded, err := url.PathUnescape(raw)
	if err != nil {
		decoded = raw
	}
	if decoded == "" {
		decoded = "/"
	}
	return URIEncode(decoded, false)
}

// CanonicalQueryString builds the sorted, URI-encoded query (slash → %2F).
func CanonicalQueryString(u *url.URL) string {
	return CanonicalQueryFromValues(u.Query())
}

// CanonicalQueryFromValues is the testable form of query canonicalization.
func CanonicalQueryFromValues(q url.Values) string {
	if len(q) == 0 {
		return ""
	}
	type pair struct{ name, value string }
	var pairs []pair
	for name, values := range q {
		encName := URIEncode(name, true)
		if len(values) == 0 {
			pairs = append(pairs, pair{encName, ""})
			continue
		}
		for _, v := range values {
			pairs = append(pairs, pair{encName, URIEncode(v, true)})
		}
	}
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].name == pairs[j].name {
			return pairs[i].value < pairs[j].value
		}
		return pairs[i].name < pairs[j].name
	})
	parts := make([]string, len(pairs))
	for i, p := range pairs {
		parts[i] = p.name + "=" + p.value
	}
	return strings.Join(parts, "&")
}

// URIEncode percent-encodes per AWS SigV4 (RFC 3986 unreserved left alone).
func URIEncode(s string, encodeSlash bool) string {
	var b strings.Builder
	b.Grow(len(s))
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') ||
			c == '-' || c == '.' || c == '_' || c == '~':
			b.WriteByte(c)
		case c == '/':
			if encodeSlash {
				b.WriteString("%2F")
			} else {
				b.WriteByte('/')
			}
		default:
			fmt.Fprintf(&b, "%%%02X", c)
		}
	}
	return b.String()
}

func deriveKey(secret, dateStamp, region, service string) []byte {
	kDate := hmacSHA256([]byte("AWS4"+secret), []byte(dateStamp))
	kRegion := hmacSHA256(kDate, []byte(region))
	kService := hmacSHA256(kRegion, []byte(service))
	return hmacSHA256(kService, []byte("aws4_request"))
}

func hmacSHA256(key, data []byte) []byte {
	m := hmac.New(sha256.New, key)
	m.Write(data)
	return m.Sum(nil)
}

func hexSHA256(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

// PayloadSHA256 returns the hex SHA256 of data.
func PayloadSHA256(data []byte) string {
	return hexSHA256(data)
}

File: windows/src/CryptoMako.S3/SigV4.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;

namespace CryptoMako.S3;

/// <summary>AWS Signature Version 4 for S3 (mirrors Sources/CryptoMakoS3/SigV4.swift).</summary>
public static class SigV4
{
    public const string EmptyPayloadSha256 =
        "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";

    public sealed class Credentials
    {
        public required string AccessKey { get; init; }
        public required string SecretKey { get; init; }
        public required string Region { get; init; }
        public string Service { get; init; } = "s3";
    }

    /// <summary>Returns headers to attach (host, x-amz-date, x-amz-content-sha256, Authorization).</summary>
    public static Dictionary<string, string> Sign(
        HttpMethod method,
        Uri url,
        Credentials credentials,
        string payloadHash = EmptyPayloadSha256,
        DateTimeOffset? now = null)
    {
        var instant = now ?? DateTimeOffset.UtcNow;
        var amzDate = instant.UtcDateTime.ToString("yyyyMMdd'T'HHmmss'Z'", CultureInfo.InvariantCulture);
        var dateStamp = amzDate[..8];

        var hostHeader = url.IsDefaultPort
            ? url.Host
            : $"{url.Host}:{url.Port}";

        var headers = new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["host"] = hostHeader,
            ["x-amz-date"] = amzDate,
            ["x-amz-content-sha256"] = payloadHash,
        };

        var canonicalUri = CanonicalPath(url);
        var canonicalQuery = CanonicalQueryString(url);
        var sortedKeys = headers.Keys.OrderBy(k => k, StringComparer.Ordinal).ToList();
        var canonicalHeaders = string.Concat(sortedKeys.Select(k => $"{k}:{headers[k].Trim()}\n"));
        var signedHeaders = string.Join(';', sortedKeys);

        var canonicalRequest = string.Join('\n',
            method.Method,
            canonicalUri,
            canonicalQuery,
            canonicalHeaders,
            signedHeaders,
            payloadHash);

        var scope = $"{dateStamp}/{credentials.Region}/{credentials.Service}/aws4_request";
        var stringToSign = string.Join('\n',
            "AWS4-HMAC-SHA256",
            amzDate,
            scope,
            Hex(SHA256.HashData(Encoding.UTF8.GetBytes(canonicalRequest))));

        var signingKey = DerivedKey(credentials.SecretKey, dateStamp, credentials.Region, credentials.Service);
        try
        {
            var signature = Hex(HmacSha256(signingKey, Encoding.UTF8.GetBytes(stringToSign)));

            headers["Authorization"] =
                "AWS4-HMAC-SHA256 " +
                $"Credential={credentials.AccessKey}/{scope}, " +
                $"SignedHeaders={signedHeaders}, " +
                $"Signature={signature}";

            return headers;
        }
        finally
        {
            CryptographicOperations.ZeroMemory(signingKey);
        }
    }

    public static string CanonicalPath(Uri url)
    {
        // Use AbsolutePath so trailing slash is preserved (ListObjectsV2 on /bucket/).
        var raw = url.AbsolutePath;
        if (string.IsNullOrEmpty(raw))
            raw = "/";
        var decoded = Uri.UnescapeDataString(raw);
        return UriEncode(decoded, encodeSlash: false);
    }

    public static string CanonicalQueryString(Uri url)
    {
        var query = url.Query;
        if (string.IsNullOrEmpty(query) || query == "?")
            return "";
        var items = ParseQuery(query.TrimStart('?'));
        return CanonicalQueryString(items);
    }

    public static string CanonicalQueryString(IEnumerable<KeyValuePair<string, string>> items)
    {
        var pairs = items
            .Select(kv => (
                Name: UriEncode(kv.Key, encodeSlash: true),
                Value: UriEncode(kv.Value ?? "", encodeSlash: true)))
            .OrderBy(p => p.Name, StringComparer.Ordinal)
            .ThenBy(p => p.Value, StringComparer.Ordinal)
            .Select(p => p.Name + "=" + p.Value);
        return string.Join('&', pairs);
    }

    public static string UriEncode(string value, bool encodeSlash)
    {
        var sb = new StringBuilder(value.Length * 2);
        foreach (var b in Encoding.UTF8.GetBytes(value))
        {
            if ((b >= 'A' && b <= 'Z') || (b >= 'a' && b <= 'z') || (b >= '0' && b <= '9')
                || b is (byte)'-' or (byte)'.' or (byte)'_' or (byte)'~')
            {
                sb.Append((char)b);
            }
            else if (b == (byte)'/' && !encodeSlash)
            {
                sb.Append('/');
            }
            else
            {
                sb.Append('%');
                sb.Append(b.ToString("X2", CultureInfo.InvariantCulture));
            }
        }
        return sb.ToString();
    }

    public static string Hex(ReadOnlySpan<byte> bytes)
    {
        var sb = new StringBuilder(bytes.Length * 2);
        foreach (var b in bytes)
            sb.Append(b.ToString("x2", CultureInfo.InvariantCulture));
        return sb.ToString();
    }

    public static byte[] DerivedKey(string secret, string dateStamp, string region, string service)
    {
        // Zero each intermediate HMAC key; caller must ZeroMemory the returned signing key.
        var key = Encoding.UTF8.GetBytes("AWS4" + secret);
        try
        {
            var next = HmacSha256(key, Encoding.UTF8.GetBytes(dateStamp));
            CryptographicOperations.ZeroMemory(key);
            key = next;
            next = HmacSha256(key, Encoding.UTF8.GetBytes(region));
            CryptographicOperations.ZeroMemory(key);
            key = next;
            next = HmacSha256(key, Encoding.UTF8.GetBytes(service));
            CryptographicOperations.ZeroMemory(key);
            key = next;
            next = HmacSha256(key, Encoding.UTF8.GetBytes("aws4_request"));
            CryptographicOperations.ZeroMemory(key);
            key = next;
            return key;
        }
        catch
        {
            CryptographicOperations.ZeroMemory(key);
            throw;
        }
    }

    private static byte[] HmacSha256(byte[] key, byte[] data)
    {
        using var hmac = new HMACSHA256(key);
        return hmac.ComputeHash(data);
    }

    private static List<KeyValuePair<string, string>> ParseQuery(string query)
    {
        var list = new List<KeyValuePair<string, string>>();
        foreach (var part in query.Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var eq = part.IndexOf('=');
            if (eq < 0)
                list.Add(new(Uri.UnescapeDataString(part), ""));
            else
                list.Add(new(
                    Uri.UnescapeDataString(part[..eq]),
                    Uri.UnescapeDataString(part[(eq + 1)..])));
        }
        return list;
    }
}
